include/unifex/stop_when.hpp

```diff
@@ -0,0 +1,429 @@
+        return op_->stopSource_.get_token();
+      }
+
+      const Receiver& get_receiver() const noexcept { return op_->receiver_; }
+
+      operation_state* op_;
+    };
+
+    template <typename Source, typename Trigger, typename Receiver>
+    class _op<Source, Trigger, Receiver>::type {
+      using source_receiver =
+          stop_when_source_receiver<Source, Trigger, Receiver>;
+      using trigger_receiver =
+          stop_when_trigger_receiver<Source, Trigger, Receiver>;
+
+    public:
+      template <typename Receiver2>
+      explicit type(
+          Source&& source,
+          Trigger&& trigger,
+          Receiver2&&
+              receiver) noexcept(is_nothrow_connectable_v<Source, source_receiver>&&
+                                     is_nothrow_connectable_v<
+                                         Trigger,
+                                         trigger_receiver>&&
+                                         std::is_nothrow_constructible_v<
+                                             Receiver,
+                                             Receiver2>)
+        : receiver_((Receiver2 &&) receiver)
+        , sourceOp_(unifex::connect((Source &&) source, source_receiver{this}))
+        , triggerOp_(
+              unifex::connect((Trigger &&) trigger, trigger_receiver{this})) {}
+
+      void start() & noexcept {
+        stopCallback_.emplace(get_stop_token(receiver_), cancel_callback{this});
+
+        unifex::start(sourceOp_);
+        unifex::start(triggerOp_);
+      }
+
+    private:
+      friend class _srcvr<Source, Trigger, Receiver>::type;
+      friend class _trcvr<Source, Trigger, Receiver>::type;
+
+      class cancel_callback {
+      public:
+        explicit cancel_callback(type* op) noexcept : op_(op) {}
+
+        void operator()() noexcept { op_->stopSource_.request_stop(); }
+
+      private:
+        type* op_;
+      };
+
+      void notify_source_complete() noexcept {
+        this->notify_trigger_complete();
+      }
+
+      void notify_trigger_complete() noexcept {
+        stopSource_.request_stop();
+        if (activeOpCount_.fetch_sub(1, std::memory_order_acq_rel) == 1) {
+          stopCallback_.reset();
+          deliver_result();
+        }
+      }
+
+      void deliver_result() noexcept {
+        try {
+          std::visit(
+              [this](auto&& tuple) {
+                if constexpr (
+                    std::tuple_size_v<
+                        std::remove_reference_t<decltype(tuple)>> != 0) {
+                  std::apply(
+                      [&](auto set_xxx, auto&&... args) {
+                        set_xxx(
+                            std::move(receiver_),
+                            static_cast<decltype(args)>(args)...);
+                      },
+                      static_cast<decltype(tuple)>(tuple));
+                } else {
+                  // Should be unreachable
+                  std::terminate();
```

Comment:
Why not make it a `static_assert` instead of an `if constexpr` if it can't happen?

Reply:
The variant does actually contain an empty tuple<> as the first alternative (a bit like the monostate).
So this callback will be instantiated for that alternative, we just should never be hitting that overload at runtime since this function should only be called after initialising the variant to some other alternative.